rasa/core/brokers/pika.py

```diff
@@ -178,7 +192,53 @@ def from_endpoint_config(
 
         return cls(broker_config.url, **broker_config.kwargs)
 
-    def publish(self, event: Dict, retries=60, retry_delay_in_seconds: int = 5) -> None:
+    def _run_pika(self) -> None:
+        self._pika_connection = self._get_connection()
+        # Run Pika io loop in extra thread so it's not blocking
+        self._run_pika_io_loop_in_thread()
+
+    def _get_connection(self) -> "Connection":
+        import pika
+
+        parameters = _get_pika_parameters(
+            self.host, self.username, self.password, self.port
+        )
+        return pika.SelectConnection(
+            parameters,
+            on_open_callback=self._on_open_connection,
+            on_open_error_callback=self._on_open_connection_error,
+        )
+
+    def _on_open_connection(self, connection: "SelectConnection") -> None:
+        logger.debug(f"Rabbit MQ connection to '{self.host}' was established.")
+        connection.channel(on_open_callback=self._on_channel_open)
+
+    def _on_open_connection_error(self, _, error: Text) -> None:
+        logger.warning(
+            f"Connecting to '{self.host}' failed with error '{error}. Trying again."
+        )
+
+    def _on_channel_open(self, channel: "Channel") -> None:
+        logger.debug("Rabbit MQ channel was opened.")
+        channel.queue_declare(self.queue, durable=True)
+
+        self.channel = channel
+
+        while len(self._unpublished_messages) > 0:
+            # Send unpublished messages
+            message = self._unpublished_messages.pop()
```

Comment:
```suggestion
            message = self._unpublished_messages.pop(0)
```
don't we want to publish the oldest message first?

Reply:
but in any case I think a `collections.deque` would probably be more appropriate, with a `q.popleft()` in the loop

Reply:
how about a debug message on having removed a message from the queue, and also the current length of the queue?